src/main/java/reposense/git/GitClone.java

```diff
@@ -68,15 +79,36 @@ public static void clone(RepoConfiguration repoConfig)
     }
 
     /**
-     * Clones a repo given the repo location into a directory.
-     * @throws IOException if it fails to create a directory.
+     * Clones a bare repo specified in {@code config} into the folder {@code outputFolderName}.
+     * @throws IOException if it fails to delete a directory.
      */
-    private static void clone(RepoLocation location, String repoFolderName, String repoName, String additionalCommand)
-            throws IOException {
-        Path rootPath = Paths.get(FileUtil.REPOS_ADDRESS, repoFolderName);
+    public static void cloneBare(RepoConfiguration config, String outputFolderName) throws IOException {
+        Path rootPath = Paths.get(FileUtil.REPOS_ADDRESS, config.getRepoFolderName());
+        FileUtil.deleteDirectory(Paths.get(rootPath.toString(), outputFolderName).toString());
         Files.createDirectories(rootPath);
-        String command =
-                String.format("git clone %s %s %s", additionalCommand, addQuote(location.toString()), repoName);
+        String command = getCloneBareCommand(config, outputFolderName);
+        runCommand(rootPath, command);
+    }
+
+    /**
+     * Performs a full clone from {@code clonedBareRepoLocation} into the folder {@code outputFolderName} and
+     * directly branches out to {@code targetBranch}.
+     * @throws IOException if it fails to delete a directory.
+     */
+    public static void cloneFromBareAndUpdateBranch(Path rootPath, Path clonedBareRepoLocation,
```

Comment:
This method is only call only
- when repo is cloned from the remote successfully
- the branch(es) exist
- no illegal file names were found (in windows)

Then only this method is called. Also, the cloning is done locally.

If a runtime exception is caught it can only be due to I/O issue. Should we still throw it as I/O exception?

Reply:
IOException is not a RuntimeException. So any exception that comes from commandrunner is currently not handled. Please catch and handle them.

As for the IOException, do catch it as well, log it and throw it out as GitCloneException as all this contributes to cloning error.

Reply:
What I meant was that if running `runCommand(...)` ends up throwing a RuntimeException, it can only be due to related to file handling issues (like user were to go and delete those files when RepoSense is still analyzing), as all possible scenarios (failure to clone due to invalid URL, invalid branch name, illegal file names) have already been handled beforehand. 

But nevertheless, I'll catch the RTE and throw it as GitCloneException so as to avoid an abrupt program termination when the other repos are being analyzed.

Reply:
>What I meant was that if running runCommand(...) ends up throwing a RuntimeException, it can only be due to related to file handling issues (like user were to go and delete those files when RepoSense is still analyzing), as all possible scenarios (failure to clone due to invalid URL, invalid branch name, illegal file names) have already been handled beforehand.

Yes, there could also be bunch of other various reasons such as corrupted files etc. In the end it will still be the git clone command throwing the error, hence we should handle it as such.

Reply:
Ok noted